pkg/http/middlewares/authorization/claims.go

```diff
@@ -28,51 +30,85 @@ var (
-// 	IsTenantAdmin    bool     `protobuf:"varint,14,opt,name=IsTenantAdmin,json=isTenantAdmin,proto3" json:"IsTenantAdmin,omitempty"`
-// 	AdminRealmIDs    []string `protobuf:"bytes,15,rep,name=AdminRealmIDs,json=adminRealmIDs,proto3" json:"AdminRealmIDs,omitempty"`
-// 	AuthenticationMethodReferences []string `protobuf:"bytes,16,rep,name=AuthenticationMethodReferences,json=amr,proto3" json:"AuthenticationMethodReferences,omitempty"`
-// }
+// Claims represents the expected claims that should be in JWT claims of an X-Request-Token
 type Claims struct {
-	ID                             string    `json:"id"`
-	IssuedAt                       Timestamp `json:"iat"`
-	NotBefore                      Timestamp `json:"nbf"`
-	Expires                        Timestamp `json:"exp"`
-	Issuer                         string    `json:"iss"`
-	UserID                         string    `json:"sub"`
-	UserName                       string    `json:"name"`
-	TenantID                       string    `json:"tenantID"`
-	Email                          string    `json:"email"`
-	RealmIDs                       []string  `json:"realmIDs"`
-	GroupIDs                       []string  `json:"groupIDs"`
-	ResourceTokenIDs               []string  `json:"resourceTokenIDs"`
-	AllowedIPs                     []string  `json:"allowedIPs"`
-	IsTenantAdmin                  bool      `json:"isTenantAdmin"`
-	AdminRealmIDs                  []string  `json:"adminRealmIDs"`
-	SourceToken                    string    `json:"-"`
-	AuthenticationMethodReferences []string  `json:"amr"`
+	// standard oidc claims
+	ID        string    `json:"id"`
+	Issuer    string    `json:"iss"`
+	IssuedAt  Timestamp `json:"iat"`
+	NotBefore Timestamp `json:"nbf"`
+	Expires   Timestamp `json:"exp"`
+	Audience  string    `json:"aud,omitempty"`
+
+	UserID   string `json:"sub"`
+	UserName string `json:"name"`
+	Email    string `json:"email"`
+
+	// Contiamo specific claims
+	TenantID      string   `json:"tenantID"`
+	RealmIDs      []string `json:"realmIDs"`
+	GroupIDs      []string `json:"groupIDs"`
+	AllowedIPs    []string `json:"allowedIPs"`
+	IsTenantAdmin bool     `json:"isTenantAdmin"`
+	AdminRealmIDs []string `json:"adminRealmIDs"`
+
+	AuthenticationMethodReferences []string `json:"amr"`
+	// AuthorizedParty is used to indicate that the request is authorizing as a
+	// service request, giving it super-admin privileges to completely any request.
+	// This replaces the "project admin" behavior of the current tokens.
+	AuthorizedParty string `json:"azp,omitempty"`
+
+	// SourceToken is for internal usage only
+	SourceToken string `json:"-"`
 }
 
 // Valid tests if the Claims object contains the minimal required information
 // to be used for authorization checks.
+//
+// Deprecated: Use the Validate method to get a precise error message. This
+// method remains for backward compatibility.
 func (a *Claims) Valid() bool {
-	return a.UserID != "" || len(a.ResourceTokenIDs) > 0
+
+	return a.Validate() == nil
+}
```

Comment:
I think it does not really make sense to keep this function, especially because it collides with the JWT interface.

Reply:
this method is used very frequently in the Hub handlers, so removing it would be a big breaking change

https://github.com/contiamo/hub/blob/7a6ccf0a1dcc6651c3ab07cd39b813421577c4d7/pkg/http/handlers/datasources_handler.go#L62

Reply:
RE this change, how do you feel about removing this usage from the handler implementations? it should be fare to assume the middleware will not set an invalid set of Claims into the context. I would be willing to remove the usage from the other projects and then come back to this method and make a breaking change to implement the jwt.Claim interface instead.


:thinking: actually even the current diff might cause some test failures, since the fixture tokens only have a 1 mintue lifetime and the `claims.Valid` now more strict. But we can probably fix that by extending the test tokens to a 5 min expiration window, since the entire test suite has never taken more than a couple of minutes. We have other issues if the token is exprired by the time a test runs (at 5 mins)